### PR TITLE
Use taggle.js library to add keywords to a Contribution

### PR DESCRIPTION
**Pull request by mic4ael**

### CODE REVIEW

File: indico/htdocs/sass/partials/_jquery-indico-taglist.scss [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+.i-taglist-field {
+	@include border-radius();
```
Comment (nop33) [inline indico/htdocs/sass/partials/_jquery-indico-taglist.scss]: Dat indentation though! It's everywhere! :dancers:

File: indico/web/templates/forms/tag_list_widget.html [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+            'use strict';
+
+            new Taggle('{{ field.id }}-tags', {
+                allowDuplicates: false,
+                preserveCase: true,
+                hiddenInputName: {{ field.id|tojson }},
+                placeholder: {{ _('Add new keyword')|tojson}},
+                tags: {{ field._value()|tojson }}
+            });
+
```
Comment (nop33) [inline indico/web/templates/forms/tag_list_widget.html]: I would delete this line

File: indico/htdocs/sass/partials/_jquery-indico-taglist.scss (hunk 1 above)
Comment (mic4ael) [inline indico/htdocs/sass/partials/_jquery-indico-taglist.scss]: really weird, need to check why it is this wau

**Comment by nop33** [comment]
Tabs instead of spaces? :O

File: indico/htdocs/sass/partials/_jquery-indico-taglist.scss (hunk 1 above)
Comment (ThiefMaster) [inline indico/htdocs/sass/partials/_jquery-indico-taglist.scss]: :cancer:

File: indico/modules/events/contributions/views.py [hunk 3]
```diff
@@ -25,7 +25,8 @@ class WPManageContributions(WPJinjaMixin, WPConferenceModifBase):
     sidemenu_option = 'contributions'
 
     def getJSFiles(self):
-        return (WPConferenceModifBase.getJSFiles(self) + self._asset_env['modules_contributions_js'].urls())
+        return (WPConferenceModifBase.getJSFiles(self) + self._asset_env['modules_contributions_js'].urls() +
+                self._asset_env['taggle_js'].urls())
```
Comment (ThiefMaster) [inline indico/modules/events/contributions/views.py]: I think you need to include this on the WP used for meeting display too (or just include it in the global bundle) - you have an "edit contribution" dialog there after all

File: indico/web/assets/bundles.py [hunk 4]
```diff
@@ -255,6 +255,8 @@ def rjs_bundle(name, *files):
                        'css/lib/selectize.js/selectize.default.css',
                        filters='cssmin', output='css/selectize_css_%(version)s.min.css')
 
+taggle_js = rjs_bundle('taggle_js', 'js/lib/taggle.js/taggle.js')
```
Comment (ThiefMaster) [inline indico/web/assets/bundles.py]: if there's only one file let's just copy it to `js/lib/`?

File: fabfile.py [hunk 5]
```diff
@@ -429,6 +429,13 @@ def install_moment_js():
         local('cp locale/es.js {0}/locale/'.format(dest_dir))
 
 
+@recipe('taggle.js')
+def install_tagging_js():
+    with lcd(os.path.join(env.ext_dir, 'taggle.js')):
+        dest_js_dir = os.path.join(lib_dir(env.src_dir, 'js'), 'taggle.js')
+        local('mkdir -p {0}'.format(dest_js_dir))
```
Comment (ThiefMaster) [inline fabfile.py]: no need for the directory just to put a single file in there

File: indico/web/templates/forms/tag_list_widget.html [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+{% block javascript -%}
+    <script>
+        (function() {
+            'use strict';
+
+            new Taggle('{{ field.id }}-tags', {
+                allowDuplicates: false,
+                preserveCase: true,
+                hiddenInputName: {{ field.id|tojson }},
+                placeholder: {{ _('Add new keyword')|tojson}},
```
Comment (ThiefMaster) [inline indico/web/templates/forms/tag_list_widget.html]: `placeholder: $T.gettext('Add new keyword'),`

File: indico/web/templates/forms/tag_list_widget.html [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+{% block javascript -%}
+    <script>
+        (function() {
+            'use strict';
+
+            new Taggle('{{ field.id }}-tags', {
+                allowDuplicates: false,
+                preserveCase: true,
+                hiddenInputName: {{ field.id|tojson }},
```
Comment (ThiefMaster) [inline indico/web/templates/forms/tag_list_widget.html]: `field.name|tojson`

File: indico/htdocs/js/indico/jquery/ajaxdialog.js [hunk 8]
```diff
@@ -140,7 +139,9 @@
                 popup.canvas.on('ajaxDialog:reload', function() {
                     loadDialog();
                 });
+
                 injectJS(dialogData.js);
+                ajaxifyForms();
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/ajaxdialog.js]: what was the reason for moving this down here?

Comment (mic4ael) [inline indico/htdocs/js/indico/jquery/ajaxdialog.js]: The problem was that setting initial data on a form was executed before injecting JS code related to new widget. The library apparently is responsible for creating a hidden input, so in that case, the data from the field was not included in the form's initialData.

Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/ajaxdialog.js]: let's hope it doesn't break anything;)

**Comment by ThiefMaster** [comment]
Merged Merged dc8a911

**Pull request closed without merging**